Working Cisco IOS macro editing (fixes #55211)

ajsiersema opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Working implementation for editing Cisco IOS macros
Fixes #55211

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ios_config

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

Commit:
Working Cisco IOS macro editing (fixes #55211)

Commit:
Working Cisco IOS macro editing [fixed pylints] (fixes #55211)

pabelanger commented:
recheck

pabelanger commented:
We should be able to write a test case for this, in tests/integration/targets/ios_config.

pabelanger commented:
The ios-python27 failure is know issue, we are working to fix it.

justjais commented:
recheck

justjais commented:
@ajsiersema can u plz write a test WRT to this fix in ios integration tests before we can go ahead and merge the PR. Also, I think you'll need to rebase the branch as well for changes to go in `2.9` i.e. devel.

ajsiersema commented:
@justjais : I see the build passed testing. Are these only python tests , i.e. extra ios integration tests are still needed ?
I'll look into a 2.9 rebase

justjais commented:
@ajsiersema You'll have to write new integration tests for the respective use case in ios_config integration test and it's not python based it's similar to the ansible play you actually run with assert to test and verify the output of the play (ref: [ios_config integration tests](https://github.com/ansible/ansible/tree/devel/test/integration/targets/ios_config/tests/cli)).

Also, rebase is not needed as it's already in `devel`

justjais reviewed (commented):
@ajsiersema Needs integration tests for the added use case.

ajsiersema commented:
> @ajsiersema Needs integration tests for the added use case.

I've create integration tests, now working on getting these into the PR

Commit:
Added integration tests for edit_macro

ajsiersema commented:
Added integration tests and tested those against the latest devel. Still failed on our switches.

E.g.

**macro.j2**
```
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
```

Running ```ansible-test network-integration --color no --local --inventory ~/inventory.networking --testcase macro --tags="cli" ios_config```

results in:

```
TASK [ios_config : Define macro name]
  ok: [testswitch]
TASK [ios_config : setup - remove configuration]
  changed: [testswitch]
TASK [ios_config : configure macro]
  changed: [testswitch]
TASK [ios_config : Check if macro has been added]
ok: [testswitch] => changed=false
  msg: All assertions passed
TASK [ios_config : configure macro again - idempotent check]
  changed: [testswitch]
TASK [ios_config : macro already/correctly configured ?]
  fatal: [testswitch]: FAILED! => changed=false
    assertion: result.changed == false
    evaluated_to: false
    msg: Assertion failed
```

and a running switch config:

```
macro name MACRO_ANSIBLE_TEST

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789

config terminal
 macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
@

```


Adding some more bytes to **macro.j2** :
```
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
 H12345678 123456789 123456789 123456789 123456789 123456789 123456789
```

results in:

```
TASK [ios_config : Define macro name]
  ok: [testswitch]
TASK [ios_config : setup - remove configuration]
  changed: [testswitch]
TASK [ios_config : configure macro]
  An exception occurred during task execution. To see the full traceback, use -vvv. The error was: mySwitch(config)#
  fatal: [testswitch]: FAILED! => changed=false
    module_stderr: |-
      Traceback (most recent call last):
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 139, in <module>
          _ansiballz_main()
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 131, in _ansiballz_main
          invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
        File "/home/user/.ansible/tmp/ansible-local-2480422uf___e/ansible-tmp-1561988531.8816607-108837394218444/AnsiballZ_ios_config.py", line 68, in invoke_module
          imp.load_module('__main__', mod, module, MOD_DESC)
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 569, in <module>
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 500, in main
        File "/tmp/ansible_ios_config_payload_8umhKB/__main__.py", line 362, in edit_config_or_macro
        File "/tmp/ansible_ios_config_payload_8umhKB/ansible_ios_config_payload.zip/ansible/module_utils/connection.py", line 186, in __rpc__
      ansible.module_utils.connection.ConnectionError: 
      % Macro length exceed its limit of 3000 bytes. Truncating the macro.
      mySwitch(config)#
      mySwitch(config)#config terminal
      config terminal
             ^
      % Invalid input detected at '^' marker.
      mySwitch(config)#
    module_stdout: ''
    msg: |-
      MODULE FAILURE
      See stdout/stderr for the exact error
    rc: 1
```

and running switch config:

```
macro name MACRO_ANSIBLE_TEST
 A12345678 123456789 123456789 123456789 123456789 123456789 123456789
 B12345678 123456789 123456789 123456789 123456789 123456789 123456789
 C12345678 123456789 123456789 123456789 123456789 123456789 123456789
 D12345678 123456789 123456789 123456789 123456789 123456789 123456789
 E12345678 123456789 123456789 123456789 123456789 123456789 123456789
 F12345678 123456789 123456789 123456789 123456789 123456789 123456789
 G12345678 123456789 123456789 123456789 123456789 123456789 123456789
 H12345678 123456789 123456789 123456789 123456789 123456789 123456789
@

```

ajsiersema commented:
Obviously with the changes in this PR, the integration tests will pass :)

justjais commented:
recheck

justjais commented:
@ajsiersema Thanks for pushing the integration tests, but currently the tests are failing coz of: https://logs.zuul.ansible.com/20/55220/de32d995cd7394e72b62127e19cc6d73cc2ffed7/third-party-check/ansible-test-network-integration-ios-python37/b5284bd/controller/ara-report/file/02c40892-71db-4e33-81f6-6f56703ccde9/#line-9 (ref: https://logs.zuul.ansible.com/20/55220/de32d995cd7394e72b62127e19cc6d73cc2ffed7/third-party-check/ansible-test-network-integration-ios-python37/b5284bd/controller/ara-report/).
Kindly, resolve the error before we can merge the PR.

hexdump0x0200 commented:
@ajsiersema, @justjais Error in integration tests looks as if the device on which they run does not support macros. This is quite possible even for new devices. I have some quite modern devices that do not support macros. There are even two devices of different hardware platforms running on the same IOS version, with one device supporting macros, and the second is not.

The proposed PR works on my equipment and integration tests are also completed successfully.

justjais commented:
@hexdump0x0200 can you plz let me know the version of ios on which you are able to run the respective integration test and version on which it fails.

ajsiersema commented:
> @hexdump0x0200 can you plz let me know the version of ios on which you are able to run the 
> respective integration test and version on which it fails.

I successfully ran the integration tests on these switches and software versions :
9407R IOS XE Version 16.9.03
4503E IOS XE Version 03.06.08.E.152-2.E8
3560CG IOS Version 15.0(2)SE12
3750X IOS Version 15.0(2)SE11

ajsiersema commented:
Do we need a conditional check in the integration test playbook ?

If the PR + checks succeed, do I need to create changelog fragments and backports to e.g. 2,8 for inclusion or will the ansible networking team take care of getting the code into stable releases ?

justjais commented:
@ajsiersema Yes, since lab IOS version is `15.6(2)T` it's better to have a conditional statement to run respective tests for the macro supported IOS version.

Also, once this PR is merged in `devel` then u can create a backport for the fix under `stable-2.8` with changelog for the backport PR.

Commit:
Added conditional integration tests for ios_config macro

ajsiersema commented:
I've added a check for macro support. This will however skip all edit_macro() integration checks in the current lab environment. Not quite sure if that should be the end result. I'd rather have an IOS lab environment on which these integration tests are automatically running as well.

hexdump0x0200 commented:
For now it looks like a good solution. Macros are most in demand on the switches, it seems that the vendor is not in a hurry to implement this function on routers (integration tests are performed on a virtual router). However, I found an old router with IOS 15.1 which supports macros, so maybe it's a matter of time when support appears in branch 15.6.

hexdump0x0200 commented:
The "when" condition for the block at the very end is very poorly read. It’s not even immediately clear to what it refers: to the block or to the last task from the block, one has to look closely at the indents.
So it would be much better:
```yaml
- name: "ios_config macro integration tests"
  when: supports_macro is succeeded
  block:
...
```

ajsiersema commented:
> 
> 
> The "when" condition for the block at the very end is very poorly read. It’s not even immediately clear to what it refers: to the block or to the last task from the block, one has to look closely at the indents.
> So it would be much better:
> 
> ```yaml
> - name: "ios_config macro integration tests"
>   when: supports_macro is succeeded
>   block:
> ...tasks
> # end block "ios_config macro integration tests"
> - name: after-block task
> ...
> ```

I check this file for the coding style:
test/integration/targets/ios_smoke/tests/cli/common_config.yaml

However, I fully agree with you with regard to the readability, my own playbooks also have the when at the top :) Will resubmit with **when** moved near **block**

Commit:
Cosmetic fix: when near block for added conditional integration tests for ios_config macro

jamiereid commented:
Hi Everyone,

I've hit the same error message in #55211 (@ajsiersema interestingly enough, the macro I am trying to create is also for DOT1X)

Is there some work I could maybe do to help get this PR over the line?

ajsiersema commented:
> I've hit the same error message in #55211 (@ajsiersema interestingly enough, the macro I am trying to create is also for DOT1X)
> Is there some work I could maybe do to help get this PR over the line?

Thanks for the feedback.
As soon as the PR is accepted, I don't mind getting submitting the PR for 2.8 / 2.9. First, someone from the Ansible network team needs to accept the PR.
@hexdump0x0200  , @justjais  : is there anything else we can/need to do ?

justjais commented on test/integration/targets/ios_config/tests/cli/macro.yaml:
```diff
@@ -0,0 +1,65 @@
+      that:
+        - "result.changed == false"
+                    
+  - name: "teardown"
```
@ajsiersema Can u move the teardown play to `always` block, which will ensure that even if the play fails in between teardown will run always.
```suggestion
  always:
  - name: "teardown"
```

ajsiersema replied:
@justjais : thanks for the review, just pushed the change (I was bogged down myself for a few days, common cold, busy schedule, etc)

justjais reviewed (commented):
@ajsiersema apologies for the delay in review, somehow this slipped my track. 
PR changes lgtm, but I have commented for a minor cosmetic update in the integration test.

Commit:
Move teardown to always for block
